crates/pallet-domains/src/staking.rs

```diff
@@ -387,119 +559,365 @@ pub(crate) fn do_deregister_operator<T: Config>(
+                if let Some(withdrawal) = withdrawals.back_mut() && withdrawal.allowed_since_domain_epoch == domain_current_epoch {
+                    withdrawal.shares = withdrawal.shares.checked_add(&shares_withdrew).ok_or(Error::ShareOverflow)?;
+                } else {
+                    withdrawals.push_back(Withdrawal {
+                        allowed_since_domain_epoch: domain_current_epoch,
+                        shares: shares_withdrew,
+                    });
                 }
 
-                let domain_stake_summary =
-                    DomainStakingSummary::<T>::get(operator.current_domain_id)
-                        .ok_or(Error::DomainNotInitialized)?;
+                *maybe_withdrawals = Some(withdrawals);
+                Ok(())
+            })
+        })
+    })
+}
 
-                let total_stake = match domain_stake_summary.current_epoch_rewards.get(&operator_id)
-                {
-                    None => operator.current_total_stake,
-                    Some(rewards) => {
-                        let operator_tax = operator.nomination_tax.mul_floor(*rewards);
-                        operator
-                            .current_total_stake
-                            .checked_add(rewards)
-                            .ok_or(Error::BalanceOverflow)?
-                            // deduct operator tax
-                            .checked_sub(&operator_tax)
-                            .ok_or(Error::BalanceUnderflow)?
-                    }
-                };
+/// Unlocks any withdraws that are ready to be unlocked.
+pub(crate) fn do_unlock_funds<T: Config>(
+    operator_id: OperatorId,
+    nominator_id: NominatorId<T>,
+) -> Result<BalanceOf<T>, Error> {
+    Withdrawals::<T>::try_mutate_exists(operator_id, nominator_id.clone(), |maybe_withdrawals| {
+        let withdrawals = maybe_withdrawals.as_mut().ok_or(Error::MissingWithdrawal)?;
+        let Withdrawal {
+            allowed_since_domain_epoch,
+            shares,
+        } = withdrawals.pop_front().ok_or(Error::MissingWithdrawal)?;
+
+        // if there are no withdrawals, then delete the storage as well
+        if withdrawals.is_empty() {
+            *maybe_withdrawals = None
+        }
 
-                let nominator_share =
-                    Perbill::from_rational(nominator.shares, operator.total_shares);
+        let (domain_id, epoch_idx) = allowed_since_domain_epoch.deconstruct();
 
-                let nominator_staked_amount = nominator_share.mul_floor(total_stake);
+        let epoch_completed_block =
+            DomainEpochCompleteAt::<T>::get(domain_id, epoch_idx).ok_or(Error::EpochNotComplete)?;
 
-                let nominator_remaining_amount = nominator_staked_amount
-                    .checked_sub(&withdraw_amount)
-                    .ok_or(Error::BalanceUnderflow)?;
+        let unlock_block_number = epoch_completed_block
```

Comment:
So all the withdrawals from a particular epoch are "unlocked" at the same block `epoch_completed_block+locking period`? Was it always like that?

Reply:
Ahh good point. I think I somehow missed this and unintended change.
Actually that is not the case with previous version. If an nominator unlocks at block #x then the funds are automatically unlocked at #{x + locking period}. 

With this change, like you have mentioned, all the withdrawals unlock at the same block {epoch_completed_block + locking period}. This approach kinda seems more true to what an epoch is within staking where all the withdrawals, no matter wihich block within the given epoch, are unlocked after epoch+locking period.

I do not have a strong opinion on either of the approaches since both approaches ensure withdrawl is only unlocked after unlocking period. 

Do you have any specific concerns/thoughts on either of the approaches

Reply:
I looked at the spec again and looks like we did not expand on when the unlock is open. So I kinda interpreted with what I have implemented. I think we need to expand that once we make a decision here 👍🏼

Reply:
I imagine that we will probably have frontends/wallets that claim the unlock automatically without user input after the unlock time. With this they will all claim at the same time basically, instead of it being spread-out over different blocks. Could this be a bottleneck, a spike of txs?

Reply:
yeah, if this is automated, then there would be a spike in transactions yes but I'm not sure if this would be a bottleneck though.

Reply:
@dariolina Now I remember why this is changed from the previous version. This is done specifically to handle the case of combining withdrawals that was initiated in the same epoch. We mentioned in this in spec here - https://www.notion.so/subspacelabs/Staking-v1-1-Design-Document-46dee50fc42644a6840c57e012e1055f?pvs=4#76ed30e0020544d698542d80cccd3e2d

So we were to track the individual withdrawals with block number, there there will be multiple withdrawals even with in the same epoch since unlocking period for each withdrawls  is different

Reply:
Okay, so all withdrawals of the same user in the same epoch are going to be unlocked at the same time. But still that does not mean that withdrawals of _different_ users should all be unlocked at the same time. Even if we aggregate per-user withdrawals of the same epoch their unlock time can be set to block of latest `withdrawal + locking period` updated when withdrawas are combined so we shouldn't need the epoch end.

Reply:
If what I propose allows to remove `DomainEpochCompleteAt` item IMO it's worth it. I can change the spec for the `Withdrawal` struct to have `allowed_since_block`(consensus  block?) instead of epoch and describe how it's updated.

Reply:
@dariolina if so then withdrawal should be updated to following
```
struct Withdrawal {
	domain_id: DomainId,
    allowed_since_epoch: Epoch,
	domain_block_number: BlockNumber
    shares: Share,
}
```

We use domain block since that drives the unlocking period and we still need the epoch to know wheter withdrawal was initiated within this epoch if we are planning to not have the `DomainEpochCompleteAt` IIUC

Reply:
Updated in spec ([struct](https://www.notion.so/subspacelabs/Domains-v2-Specification-3fb0ec6e4d204c4881a7df50ef58da8f?pvs=4#49aa812c04bc4c64b7144de66bbf0468), [withdrawal protocol](https://www.notion.so/subspacelabs/Domains-v2-Specification-3fb0ec6e4d204c4881a7df50ef58da8f?pvs=4#259fbadac5d14ba3892907531e5e66cb))